tests/testsuite/package.rs

```diff
@@ -3132,3 +3133,94 @@ See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for
         &[],
     );
 }
+
+#[cargo_test]
+fn include_files_called_target_project() {
+    // https://github.com/rust-lang/cargo/issues/12790
+    // files and folders called "target" should be included, unless they're the actual target directory
+    let p = init_and_add_inner_target(project())
+        .file("target/foo.txt", "")
+        .build();
+
+    p.cargo("package -l")
+        .with_stdout(
+            "\
+Cargo.lock
+Cargo.toml
+Cargo.toml.orig
+data/not_target
+data/target
+derp/not_target/foo.txt
+derp/target/foo.txt
+src/main.rs
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn include_files_called_target_git() {
+    // https://github.com/rust-lang/cargo/issues/12790
+    // files and folders called "target" should be included, unless they're the actual target directory
+    let (p, repo) = git::new_repo("target_uncommitted", |p| init_and_add_inner_target(p));
+    // add target folder but not committed.
+    let _ = fs::create_dir_all(&repo.workdir().unwrap().join("target/foo.txt")).unwrap();
+
+    p.cargo("package -l")
+        .with_stdout(
+            "\
+.cargo_vcs_info.json
+Cargo.lock
+Cargo.toml
+Cargo.toml.orig
+data/not_target
+data/target
+derp/not_target/foo.txt
+derp/target/foo.txt
+src/main.rs
+",
+        )
+        .run();
+
+    // if target is committed, it should be include.
+    let p = git::new("target_committed", |p| {
+        init_and_add_inner_target(p).file("target/foo.txt", "")
+    });
```

Comment:
Would be better if we can reuse the git repo instead of creating a new one.

```suggestion
    // if target is committed, it should be include.
    let mut index = t!(repo.index());
    t!(index.add_path(Path::new("target/foo.txt")));
    t!(index.write());
    git::commit(&repo);

```

And you might need to change code on line 3167 accordling to create a real foo.txt file:

```rust
    // add target folder but not committed.
    _ = fs::create_dir_all(p.build_dir()).unwrap();
    _ = fs::write(p.build_dir().join("foo.txt"), "").unwrap();
```

Reply:
Okay. I can see they are not equivalent, but feel that reusing the git repo make the intent a bit clear.

Reply:
> Okay. I can see they are not equivalent, but feel that reusing the git repo make the intent a bit clear.

I had reused the repo to commit the file, it looks more tidy, thanks.